Add Agent and agent run metadata and expose it on result objects and span attributes

sstanovnik opened the pull request:
This allows users to add extra metadata to the agent's span, either as a direct string or dict, or a callable that takes the RunContext. The attributes are added/computed after the agent finishes. The metadata is under the `metadata` attribute.

Closes https://github.com/pydantic/pydantic-ai/issues/3150

DouweM commented on docs/logfire.md:
```diff
@@ -356,3 +356,39 @@ Agent.instrument_all(instrumentation_settings)
+from pydantic_ai.models.instrumented import InstrumentationSettings
+
+
+def span_attribute_callback(ctx: RunContext[None]) -> dict[str, str]:
```
@sstanovnik I just mean that the metadata dict keys will not directly become span attributes, as they'd be nested under `metadata`, but if that's fine for your use case, feel free to update this PR (with force push possibly), or create a new one!

sstanovnik commented on pydantic_ai_slim/pydantic_ai/agent/__init__.py:
```diff
@@ -89,6 +90,8 @@
 S = TypeVar('S')
 NoneType = type(None)
 
+AgentMetadataValue = str | dict[str, str] | Callable[[RunContext[AgentDepsT]], str | dict[str, str]]
```
I opted to not introduce a generic `MetadataT` to the Agent, but instead allow strings or dicts, that get serialized into JSON-strings for the OTel attribute.

sstanovnik commented on pydantic_ai_slim/pydantic_ai/agent/__init__.py:
```diff
@@ -645,6 +660,7 @@ async def get_instructions(run_context: RunContext[AgentDepsT]) -> str | None:
             },
         )
 
+        run_metadata: str | dict[str, str] | None = None
```
We could technically set a non-callable metadata even if the run fails, and the ctx isn't available. What do you think? I opted for the simpler implementation.

sstanovnik commented on pydantic_ai_slim/pydantic_ai/agent/__init__.py:
```diff
@@ -292,6 +303,7 @@ def __init__(
 
         self._output_type = output_type
         self.instrument = instrument
+        self._metadata = metadata
```
Not sure about the sunder, and that it's not represented in repr. I went by feeling.

sstanovnik commented:
Apologies for the delay in responding to your review. Thank you for taking the time for this.

I believe I've addressed all comments:

- The metadata type is changed from `str | dict[str, str]` to `dict[str, Any]`.
- `logfire.agent.metadata` is now `agent.metadata`.
- Metadata is now available on run results.
- Metadata is fully computed even if the agent run fails.
- Both `run` and `iter` now have the metadata attribute and work with `.override()`.

DouweM commented on docs/agents.md:
```diff
@@ -414,6 +414,7 @@ _(This example is complete, it can be run "as is" — you'll need to add `asynci
 You can retrieve usage statistics (tokens, requests, etc.) at any time from the [`AgentRun`][pydantic_ai.agent.AgentRun] object via `agent_run.usage()`. This method returns a [`RunUsage`][pydantic_ai.usage.RunUsage] object containing the usage data.
 
 Once the run finishes, `agent_run.result` becomes a [`AgentRunResult`][pydantic_ai.agent.AgentRunResult] object containing the final output (and related metadata).
+You can inspect [`agent_run.metadata`][pydantic_ai.agent.AgentRun] or [`agent_run.result.metadata`][pydantic_ai.agent.AgentRunResult] after the run completes to read any metadata configured on the agent.
```
This deserves a separate section that also explains how to set it. I think much of what's currently in `logfire.md` can be moved here, and then those docs can link here.

sstanovnik replied:
I put most of the documentation here and only mentioned the relevant parts in `logfire.md`, linking here for details.

DouweM commented on pydantic_ai_slim/pydantic_ai/agent/__init__.py:
```diff
@@ -539,6 +561,10 @@ async def main():
             usage: Optional usage to start with, useful for resuming a conversation or agents used in tools.
+            metadata: Optional metadata to attach to this run. Accepts a dictionary or a callable taking
+                [`RunContext`][pydantic_ai.tools.RunContext]. The resolved dictionary is shallow merged into the
+                agent's metadata (or any [`Agent.override`][pydantic_ai.agent.Agent.override]) with run-level keys
```
Existing `agent.override`s always cause the `agent.run` value to be ignored, fully overwriting the agent + agent run original. We should have the same behavior here

DouweM commented on pydantic_ai_slim/pydantic_ai/agent/__init__.py:
```diff
@@ -679,18 +719,51 @@ async def get_instructions(run_context: RunContext[AgentDepsT]) -> str | None:
             finally:
                 run_span.end()
 
+    def _compute_agent_metadata(
```
For consistency with the other fields supported by `agent.override`, let's call this `_get_metadata()` and the `run_metadata_config` `additional_metadata`

DouweM commented on pydantic_ai_slim/pydantic_ai/agent/__init__.py:
```diff
@@ -679,18 +719,51 @@ async def get_instructions(run_context: RunContext[AgentDepsT]) -> str | None:
+            base_config = self._metadata
+
+        base_metadata = self._resolve_metadata_config(base_config, ctx)
+        run_metadata = self._resolve_metadata_config(run_metadata_config, ctx)
```
As mentioned above, overridden metadata fully overrides run metadata

DouweM commented on pydantic_ai_slim/pydantic_ai/agent/abstract.py:
```diff
@@ -1035,6 +1070,10 @@ async def main():
             usage: Optional usage to start with, useful for resuming a conversation or agents used in tools.
+            metadata: Optional metadata to attach to this run. Accepts a dictionary or a callable taking
+                [`RunContext`][pydantic_ai.tools.RunContext]. The resolved dictionary is shallow merged into the
+                agent's metadata (or any [`Agent.override`][pydantic_ai.agent.Agent.override]) with run-level keys
```
See above, this behavior is inconsistent with existing expectations, so the behavior and docstring will need updating

DouweM commented on pydantic_ai_slim/pydantic_ai/result.py:
```diff
@@ -126,6 +127,13 @@ def run_id(self) -> str:
+    def metadata(self) -> dict[str, Any] | None:
+        """Metadata associated with this agent run, if configured."""
+        if self._metadata_getter is not None:
+            return self._metadata_getter()
```
I understand that we do this because the metadata won't be available on `self._run_ctx` as it's only built at the end of the run, but I wonder if we can change that and simplify this.

Can we make sure the metadata is always available, just initially built off the original `RunContext` and then regenerated at the end of the run? Usually it'll just use `deps` anyway, which is available from the start.

sstanovnik commented on pydantic_ai_slim/pydantic_ai/agent/__init__.py:
```diff
@@ -724,6 +797,9 @@ def _run_span_end_attributes(
                 attrs['pydantic_ai.variable_instructions'] = True
 
+        if metadata is not None:
+            attrs['agent.metadata'] = json.dumps(metadata)
```
Done!

DouweM commented on pydantic_ai_slim/pydantic_ai/agent/__init__.py:
```diff
@@ -664,9 +691,19 @@ async def get_instructions(run_context: RunContext[AgentDepsT]) -> str | None:
+
+                    final_result = agent_run.result
+                    if instrumentation_settings and run_span.is_recording():
+                        if instrumentation_settings.include_content and final_result is not None:
```
Let's combine these 2 ifs

DouweM commented on docs/agents.md:
```diff
@@ -413,7 +413,40 @@ _(This example is complete, it can be run "as is" — you'll need to add `asynci
 You can retrieve usage statistics (tokens, requests, etc.) at any time from the [`AgentRun`][pydantic_ai.agent.AgentRun] object via `agent_run.usage()`. This method returns a [`RunUsage`][pydantic_ai.usage.RunUsage] object containing the usage data.
 
-Once the run finishes, `agent_run.result` becomes a [`AgentRunResult`][pydantic_ai.agent.AgentRunResult] object containing the final output (and related metadata).
+Once the run finishes, `agent_run.result` becomes an [`AgentRunResult`][pydantic_ai.agent.AgentRunResult] object containing the final output.
```
Let's not drop the final clause here:

```suggestion
Once the run finishes, `agent_run.result` becomes an [`AgentRunResult`][pydantic_ai.agent.AgentRunResult] object containing the final output (and related metadata).
```

DouweM commented on docs/agents.md:
```diff
@@ -413,7 +413,40 @@ _(This example is complete, it can be run "as is" — you'll need to add `asynci
-Once the run finishes, `agent_run.result` becomes a [`AgentRunResult`][pydantic_ai.agent.AgentRunResult] object containing the final output (and related metadata).
+Once the run finishes, `agent_run.result` becomes an [`AgentRunResult`][pydantic_ai.agent.AgentRunResult] object containing the final output.
+
+#### Run metadata
```
Let's move this into the `Additional Configuration` section (with the appropriate subheading level); right now in the sidebar it looks like this is specific to `Iterating Over an Agent's Graph`

DouweM commented on docs/agents.md:
```diff
@@ -413,7 +413,40 @@ _(This example is complete, it can be run "as is" — you'll need to add `asynci
+
+#### Run metadata
+
+Metadata configured on an [`Agent`][pydantic_ai.agent.Agent] or passed to a run is available after completion via [`AgentRun.metadata`][pydantic_ai.agent.AgentRun], [`AgentRunResult.metadata`][pydantic_ai.agent.AgentRunResult], or [`StreamedRunResult.metadata`][pydantic_ai.result.StreamedRunResult].
```
Let's start by introducing the feature, with an example of when it could be useful, and then explain how to set it and access it.

We should mention that it's available on `RunContext` and is set on the span attributes (which can be one of the examples of when it can be useful)

And we should mention that it can take a static dict or a function taking `RunContext`. And that it's run at the start and the end of the run.
